generic_server.cc

```diff
@@ -119,7 +119,7 @@ future<> server::stop() {
 
 future<> server::shutdown() noexcept {
     if (_gate.is_closed()) {
-        return make_ready_future<>();
+        co_return;
```

Comment:
Great catch thank you! There are a lot of pitfalls to coroutines (and something something about unexpected interactions between theoretically independent C++ features ;) ). Will fix this in v7. Thanks!

Reply:
I don't understand why anything is needed. The coroutine has an implicit try/catch around its body. Exceptions caught are sent to seastar::coroutine_traits_base::unhandled_exception(), which forwards it to the underlying promise, so it ends up in the returned future.

Reply:
IIRC, Benny's point was that an exception would prevent `c.shutdown()` from being initiated for all the connections, and then the gate would never be vacated, and `_all_connections_stopped` would never resolve. ... https://github.com/scylladb/scylladb/pull/20212#discussion_r1726742964

Reply:
I don't see how this is related to making server::shutdown() noexcept. After converting it to a coroutine, we're sure it won't throw, because coroutines package unhandled exceptions into the return future. It won't throw whether or not it's marked noexcept.

abort_accept() throwing would indeed be broken, but it doesn't (shouldn't) throw. In any case coroutinization didn't make it better or worse in that aspect.

Reply:
- assume the consumer of the `server` object calls `stop`, or `shutdown` (and in the former case, `stop` calls `shutdown`)
- `shutdown` closes the gate, `abort_accept` throws, shutdown for the connections is not initated (not reached)
- the exception is propagated to the consumer of the `server` object, which is waiting on either `stop` or `shutdown` to resolve (see first step)
- `future::get()` or `co_await` in the consumer throws; assume the consumer catches it
- what is the consumer now supposed to do? If if calls `shutdown` again, that won't do anything (the gate is closed). if the consumer calls `stop`, then the `co_await` on `_all_connections_stopped` may block indefinitely (the connections don't know they should be leaving the gate).

I believe Benny's point with making `shutdown` noexcept was to prevent this (effective) hang. If `shutdown` were to throw between `_gate.close()` and starting to shut down the connections, we'd abort the scylla process -- no hang. @bhalevy please explain better if I'm mistaken. Thanks.

I'm fine either way, BTW; I can just drop the noexcept patch, or rework it, as long as we reach consensus. Thanks.

> abort_accept() throwing would indeed be broken, but it doesn't (shouldn't) throw

Should I try marking abort_accept() noexcept then?

> In any case coroutinization didn't make it better or worse in that aspect.

I agree; it's just that coroutinization affects (somewhat) how we can express that no exception should prevent `shutdown` from reaching the `parallel_for_each`.

Reply:
done in v7